src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/IUpgradeResultCollector.cs

```diff
@@ -0,0 +1,21 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.DotNet.UpgradeAssistant.Analysis;
+
+namespace Microsoft.DotNet.UpgradeAssistant
+{
+    public interface IUpgradeResultCollector
```

Comment:
Is this needed? Couldn't we add an `ICollection<AnalyzeResultDefinition> Results { get; }` to the upgrade context?

Reply:
We can do that, but I did not want to expose all the methods like remove, clear etc to the external extensions which can unknowingly or knowingly hamper the results that will be printed.

The extensions should just be able to add their results and make them accessible to ConsoleUpgrade for writing. Ideally, I want to remove the clear method from this interface and create a way such that only ConsoleUpgrade can reset the collection once it writes the results after each step results are written. This control should not be given to the extensions.

What do you think?

Reply:
That makes sense. Mitigation for this could be that each invocation would put a new list backing in, but an uber-list would be maintained separately.

I think my push back here may also stem from it being a fairly general purpose collector, but has been constratined to be just upgrade results.  What about something like:

```csharp
public interface ICollector<T> : IEnumerable<T>
{
  void Add(T item);
  void AddRange(IEnumerable<T> items);
  void Clear(); // Do we need this if we want it to be add-only?
```

Reply:
I think letting it `: IEnumerable<T>` imbues it w/ capabilities we don't want people picking up on. This is a bag they should **only ever** populate; we take care of iteration and knowing how to iterate.

Additionally - `T` doesn't make sense because inevitably the objects all go into a writer, and a writer doesn't know how to write a `T`, it only knows how to write an `AnalyzeResultDefinition`. We could `where T : AnalyzeResultDefinition` but I'd argue for the usefulness when, again, the writer's only going to look at the properties that exist on an `AnalyzeResultDefinition`

Reply:
> T doesn't make sense because inevitably the objects all go into a writer,


Making it generic is just because it seems like something that would be reusable; i.e. a collection that can be added to, but not removed from. I'd assume the actual usage would be something like `ICollector<AnalyzeResultDefinition> IUpgradeContext.Results {get;}`

Reply:
Ah, gotchya.

Reply:
@twsouthwick can you explain how the `IUpgradeContext` would work here? I tried making it work but is seems like the `IUpgradeContext` that is passed into the steps by the `ConsoleUpgrade` command [here ](https://github.com/dotnet/upgrade-assistant/blob/e1207a2519a12504527f0cfddd4d57e0b3553abd/src/cli/Microsoft.DotNet.UpgradeAssistant.Cli/Commands/Upgrade/ConsoleUpgrade.cs#L50) is a different instance than the `IUpgradeContext` instance injected via dependency injection & obtained through constructor parameters in a step (like an Analyzer), thus they are unable to share state.

I want steps to be able to use the injected context rather than the context that is passed as parameter to the ApplyAsync method, because some steps like Roslyn Analyzers and CodeFixers do not have access to the same parameters, but do have access to the injected services.

Reply:
> a different instance than the IUpgradeContext instance injected via dependency injection

Hmmm the context is specifically *not* passed via constructor injection, but rather through parameters to methods. Initially because it required async to create, but also because state should not be something passed around via DI.

> some steps like Roslyn Analyzers and CodeFixers do not have access to the same parameters, but do have access to the injected services.

That's actually not true. They may be using them, but if they are that will be a broken experience if used standalone in VS. Analyzers should not be using them, and in VS code fixers use MEF. The analyzers/codefixers should probably not be managing any sort of report stuff. That should be metadata on top of them that gets added (i.e. there's a description field they have we can add to the report if needed).